File: apps/desktop/native/macos/Sources/PrivateAIGatewayMac/RuntimeStore.swift
import AppKit
import Foundation
import SwiftUI

@MainActor
final class RuntimeStore: ObservableObject {
    static let shared = RuntimeStore()
    @Published private(set) var state = GatewayState.loading
    @Published private(set) var agents: [AgentStatus] = []
    @Published private(set) var usage = UsagePage(
        items: [], nextCursor: nil, summary: .zero, series: [], agents: [], models: []
    )
    @Published private(set) var clientKey = ""
    @Published var errorMessage: String?
    @Published var isBusy = false
    @Published var usageAgent: String?
    @Published var usageModel: String?
    @Published var usageRange: UsageRange = .thirtyDays

    private let client = RuntimeClient()
    private var lastUsageRevision: UInt64 = 0

    init() {
        client.onState = { [weak self] state in self?.accept(state) }
        client.onExit = { [weak self] error in
            self?.errorMessage = error?.localizedDescription ?? "The desktop runtime stopped."
        }
        do {
            try client.start()
            reload()
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func shutdown() { client.shutdown() }

    var activeProfile: ConfidentialProfile? {
        state.profiles.first { $0.id == state.activeProfileId }
    }

    var isProtected: Bool {
        state.status == .verified && !state.configurationVerification
    }

    var isRunning: Bool {
        [.verifying, .verified, .blocked].contains(state.status)
    }

    var isDevMode: Bool { !state.config.requireProductionOs }

    func reload() {
        request("getState", params: EmptyParams()) { (result: Result<GatewayState, Error>) in
            if case .success(let state) = result { self.accept(state) }
        }
        reloadAgents()
        reloadUsage(reset: true)
        request("getClientKey", params: EmptyParams()) { (result: Result<String, Error>) in
            if case .success(let key) = result { self.clientKey = key }
        }
    }

    func setProtection(_ enabled: Bool) {
        guard !isBusy else { return }
        isBusy = true
        if enabled {
            request("start", params: StartParams(config: state.config)) {
                (result: Result<GatewayState, Error>) in
                self.finishState(result)
            }
        } else {
            request("stop", params: EmptyParams()) { (result: Result<GatewayState, Error>) in
                self.finishState(result)
            }
        }
    }

    func verifyAndSave(
        profile: ConfidentialProfileInput,
        allowDevOs: Bool,
        key: String?,
        completion: @escaping (Bool) -> Void
    ) {
        guard !isBusy else { return }
        isBusy = true
        request(
            "verifyConfiguration",
            params: VerifyParams(
                profile: profile,
                requireProductionOs: !allowDevOs,
                key: key?.isEmpty == false ? key : nil
            )
        ) { (result: Result<GatewayState, Error>) in
            self.isBusy = false
            switch result {
            case .success(let state):
                self.accept(state)
                completion(true)
            case .failure(let error):
                self.errorMessage = error.localizedDescription
                completion(false)
            }
        }
    }

    func activate(_ profile: ConfidentialProfile) {
        request("activateProfile", params: ProfileParams(profileId: profile.id)) {
            (result: Result<GatewayState, Error>) in self.finishState(result)
        }
    }

    func delete(_ profile: ConfidentialProfile) {
        request("deleteProfile", params: ProfileParams(profileId: profile.id)) {
            (result: Result<GatewayState, Error>) in self.finishState(result)
        }
    }

    func clearActiveCredential() {
        request("clearApiKey", params: EmptyParams()) {
            (result: Result<GatewayState, Error>) in self.finishState(result)
        }
    }

    func reloadAgents() {
        request("listAgents", params: EmptyParams()) { (result: Result<[AgentStatus], Error>) in
            switch result {
            case .success(let agents): self.agents = agents
            case .failure(let error): self.errorMessage = error.localizedDescription
            }
        }
    }

    func setAgent(_ agent: AgentStatus, connected: Bool) {
        let defaultModel = agent.id == "codex" ? state.catalog?.models.first?.id : nil
        let options = ConnectOptions(defaultModel: defaultModel)
        request(
            "previewAgent",
            params: AgentParams(agentId: agent.id, connect: connected, options: options)
        ) { (preview: Result<AgentPreview, Error>) in
            switch preview {
            case .failure(let error): self.errorMessage = error.localizedDescription
            case .success(let preview):
                self.request(
                    "applyAgent",
                    params: ApplyAgentParams(
                        agentId: agent.id,
                        connect: connected,
                        revision: preview.revision,
                        options: options
                    )
                ) { (result: Result<AgentStatus, Error>) in
                    switch result {
                    case .success: self.reloadAgents()
                    case .failure(let error): self.errorMessage = error.localizedDescription
                    }
                }
            }
        }
    }

    func restoreAllAgents() {
        request("disconnectAllAgents", params: EmptyParams()) {
            (result: Result<[AgentStatus], Error>) in
            switch result {
            case .success(let agents): self.agents = agents
            case .failure(let error): self.errorMessage = error.localizedDescription
            }
        }
    }

    func reloadUsage(reset: Bool) {
        let now = UInt64(Date().timeIntervalSince1970)
        let query = UsageQuery(
            agent: usageAgent,
            model: usageModel,
            sessionId: nil,
            since: usageRange.since(now: now),
            until: nil,
            cursor: reset ? nil : usage.nextCursor,
            limit: 20
        )
        request("queryUsage", params: UsageParams(query: query)) {
            (result: Result<UsagePage, Error>) in
            switch result {
            case .success(let page):
                if reset {
                    self.usage = page
                } else {
                    self.usage = UsagePage(
                        items: self.usage.items + page.items,
                        nextCursor: page.nextCursor,
                        summary: page.summary,
                        series: page.series,
                        agents: page.agents,
                        models: page.models
                    )
                }
            case .failure(let error): self.errorMessage = error.localizedDescription
            }
        }
    }

    func saveLocalApi(_ config: LocalApiConfig, completion: @escaping (Bool) -> Void) {
        request("saveLocalApiConfig", params: LocalApiParams(config: config)) {
            (result: Result<GatewayState, Error>) in
            switch result {
            case .success(let state): self.accept(state); completion(true)
            case .failure(let error): self.errorMessage = error.localizedDescription; completion(false)
            }
        }
    }

    func rotateClientKey() {
        request("rotateClientKey", params: EmptyParams()) { (result: Result<String, Error>) in
            switch result {
            case .success(let key): self.clientKey = key
            case .failure(let error): self.errorMessage = error.localizedDescription
            }
        }
    }

    func copy(_ value: String) {
        NSPasteboard.general.clearContents()
        NSPasteboard.general.setString(value, forType: .string)
    }

    func clearUsage() {
        request("clearUsage", params: EmptyParams()) { (result: Result<UInt64, Error>) in
            switch result {
            case .success: self.reloadUsage(reset: true)
            case .failure(let error): self.errorMessage = error.localizedDescription
            }
        }
    }

    func exportUsage(to path: String, completion: @escaping (Bool) -> Void) {
        let now = UInt64(Date().timeIntervalSince1970)
        let query = UsageQuery(
            agent: usageAgent,
            model: usageModel,
            sessionId: nil,
            since: usageRange.since(now: now),
            until: nil,
            cursor: nil,
            limit: nil
        )
        request("exportUsageCsv", params: ExportUsageParams(query: query, path: path)) {
            (result: Result<Int, Error>) in
            switch result {
            case .success: completion(true)
            case .failure(let error):
                self.errorMessage = error.localizedDescription
                completion(false)
            }
        }
    }

    private func accept(_ state: GatewayState) {
        self.state = state
        if state.usageRevision != lastUsageRevision {
            lastUsageRevision = state.usageRevision
            reloadUsage(reset: true)
        }
    }

    private func finishState(_ result: Result<GatewayState, Error>) {
        isBusy = false
        switch result {
        case .success(let state): accept(state); reloadAgents()
        case .failure(let error): errorMessage = error.localizedDescription
        }
    }

    private func request<ResultType: Decodable, Params: Encodable>(
        _ method: String,
        params: Params,
        completion: @escaping (Result<ResultType, Error>) -> Void
    ) {
        client.request(method, params: params, completion: completion)
    }
}

enum UsageRange: String, CaseIterable, Identifiable {
    case sevenDays = "7 days"
    case thirtyDays = "30 days"
    case all = "All time"
    var id: String { rawValue }
    func since(now: UInt64) -> UInt64? {
        switch self {
        case .sevenDays: now - 7 * 86_400
        case .thirtyDays: now - 30 * 86_400
        case .all: nil
        }
    }
}

File: apps/desktop/native/windows/PrivateAIGateway.Windows/RuntimeStore.cs
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace PrivateAIGateway.Windows;

public sealed class RuntimeStore : INotifyPropertyChanged, IAsyncDisposable
{
    private readonly SemaphoreSlim lifecycleLock = new(1, 1);
    private RuntimeClient? client;
    private Task cleanupTask = Task.CompletedTask;
    private GatewayState state = GatewayState.Empty;
    private AgentStatus[] agents = [];
    private UsagePage usage = UsagePage.Empty;
    private string clientKey = "";
    private bool busy;
    private bool runtimeAvailable;
    private bool disposing;

    public event PropertyChangedEventHandler? PropertyChanged;
    public event Action<string>? Error;
    public GatewayState State { get => state; private set => Set(ref state, value); }
    public AgentStatus[] Agents { get => agents; private set => Set(ref agents, value); }
    public UsagePage Usage { get => usage; private set => Set(ref usage, value); }
    public string ClientKey { get => clientKey; private set => Set(ref clientKey, value); }
    public bool IsBusy { get => busy; private set => Set(ref busy, value); }
    public bool IsRuntimeAvailable { get => runtimeAvailable; private set => Set(ref runtimeAvailable, value); }
    public string? UsageAgent { get; set; }
    public string? UsageModel { get; set; }
    public UsageRange UsageRange { get; set; } = UsageRange.ThirtyDays;
    public bool IsRunning => IsRuntimeAvailable &&
        (State.Status == "verifying" || State.Status == "verified" || State.Status == "blocked");
    public bool IsProtected => IsRuntimeAvailable && RuntimePresentation.IsProtected(State);
    public bool IsDevMode => !State.Config.RequireProductionOs;
    public ConfidentialProfile? ActiveProfile => State.Profiles.FirstOrDefault(profile => profile.Id == State.ActiveProfileId);

    public Task<bool> InitializeAsync() => RestartAsync();

    public async Task<bool> RestartAsync()
    {
        await lifecycleLock.WaitAsync();
        try
        {
            if (disposing) return false;
            IsBusy = true;
            IsRuntimeAvailable = false;
            ClearRuntimeData();
            await cleanupTask;
            cleanupTask = Task.CompletedTask;
            if (client is { } previous)
            {
                client = null;
                await previous.DisposeAsync();
            }

            var next = new RuntimeClient();
            client = next;
            next.StateChanged += state => App.MainWindow.DispatcherQueue.TryEnqueue(() => Accept(next, state));
            next.Exited += error => App.MainWindow.DispatcherQueue.TryEnqueue(() => RuntimeExited(next, error));
            try
            {
                await next.StartAsync();
                await LoadInitialDataAsync(next);
                EnsureOwned(next);
                if (!next.IsAvailable)
                    throw new RuntimeException("runtime_exited", "The desktop runtime stopped during startup.");
                IsRuntimeAvailable = true;
                return true;
            }
            catch (Exception error)
            {
                if (ReferenceEquals(client, next)) client = null;
                IsRuntimeAvailable = false;
                ClearRuntimeData();
                await next.DisposeAsync();
                ReportError(error);
                return false;
            }
        }
        finally
        {
            IsBusy = false;
            lifecycleLock.Release();
        }
    }

    public Task<GatewayState> SetProtectionAsync(bool enabled)
    {
        var runtime = CurrentClient();
        return RunStateAsync(runtime, enabled
            ? runtime.RequestAsync<GatewayState>("start", new StartParams(State.Config))
            : runtime.RequestAsync<GatewayState>("stop", new { }));
    }

    public async Task<(bool Success, string? Error)> VerifyAndSaveAsync(ConfidentialProfileInput profile, bool allowDevOs, string? key)
    {
        try
        {
            var runtime = CurrentClient();
            IsBusy = true;
            var next = await runtime.RequestAsync<GatewayState>("verifyConfiguration", new VerifyParams(
                profile, !allowDevOs, string.IsNullOrWhiteSpace(key) ? null : key));
            AcceptCurrent(runtime, next);
            return (true, null);
        }
        catch (Exception error) { return (false, error.Message); }
        finally { IsBusy = false; }
    }

    public Task<GatewayState> ActivateProfileAsync(string id)
    {
        var runtime = CurrentClient();
        return RunStateAsync(runtime, runtime.RequestAsync<GatewayState>("activateProfile", new ProfileParams(id)));
    }

    public Task<GatewayState> DeleteProfileAsync(string id)
    {
        var runtime = CurrentClient();
        return RunStateAsync(runtime, runtime.RequestAsync<GatewayState>("deleteProfile", new ProfileParams(id)));
    }

    public Task<GatewayState> ClearCredentialAsync()
    {
        var runtime = CurrentClient();
        return RunStateAsync(runtime, runtime.RequestAsync<GatewayState>("clearApiKey", new { }));
    }

    public Task<GatewayState> SaveLocalApiAsync(LocalApiConfig config)
    {
        var runtime = CurrentClient();
        return RunStateAsync(runtime, runtime.RequestAsync<GatewayState>("saveLocalApiConfig", new LocalApiParams(config)));
    }

    public async Task SetAgentAsync(AgentStatus agent, bool connected)
    {
        try
        {
            var runtime = CurrentClient();
            var defaultModel = agent.Id == "codex" ? State.Catalog?.Models.FirstOrDefault()?.Id : null;
            var options = new ConnectOptions(defaultModel);
            var preview = await runtime.RequestAsync<AgentPreview>("previewAgent", new AgentParams(agent.Id, connected, options));
            EnsureCurrent(runtime);
            await runtime.RequestAsync<AgentStatus>("applyAgent", new ApplyAgentParams(agent.Id, connected, preview.Revision, options));
            await ReloadAgentsAsync(runtime, true);
        }
        catch (Exception error) { ReportError(error); }
    }

    public async Task RestoreAllAgentsAsync()
    {
        try
        {
            var runtime = CurrentClient();
            var result = await runtime.RequestAsync<AgentStatus[]>("disconnectAllAgents", new { });
            EnsureCurrent(runtime);
            Agents = result;
        }
        catch (Exception error) { ReportError(error); }
    }

    public Task ReloadAgentsAsync() => ReloadAgentsAsync(CurrentClient(), true);
    public Task ReloadUsageAsync(bool reset) => ReloadUsageAsync(CurrentClient(), reset, true);

    public async Task ExportUsageAsync(string path)
    {
        try
        {
            var runtime = CurrentClient();
            await runtime.RequestAsync<int>("exportUsageCsv", new ExportUsageParams(CurrentUsageQuery(null, null), path));
            EnsureCurrent(runtime);
        }
        catch (Exception error) { ReportError(error); }
    }

    public async Task ClearUsageAsync()
    {
        try
        {
            var runtime = CurrentClient();
            await runtime.RequestAsync<ulong>("clearUsage", new { });
            EnsureCurrent(runtime);
            await ReloadUsageAsync(runtime, true, true);
        }
        catch (Exception error) { ReportError(error); }
    }

    public async Task RotateClientKeyAsync()
    {
        try
        {
            var runtime = CurrentClient();
            var result = await runtime.RequestAsync<string>("rotateClientKey", new { });
            EnsureCurrent(runtime);
            ClientKey = result;
        }
        catch (Exception error) { ReportError(error); }
    }

    private async Task LoadInitialDataAsync(RuntimeClient runtime)
    {
        var stateTask = runtime.RequestAsync<GatewayState>("getState", new { });
        var agentsTask = runtime.RequestAsync<AgentStatus[]>("listAgents", new { });
        var usageTask = runtime.RequestAsync<UsagePage>("queryUsage", new UsageParams(CurrentUsageQuery(null, 20)));
        var keyTask = runtime.RequestAsync<string>("getClientKey", new { });
        await Task.WhenAll(new Task[] { stateTask, agentsTask, usageTask, keyTask });
        EnsureOwned(runtime);
        State = await stateTask;
        Agents = await agentsTask;
        Usage = await usageTask;
        ClientKey = await keyTask;
        RaiseDerivedState();
    }

    private async Task ReloadAgentsAsync(RuntimeClient runtime, bool reportError)
    {
        try
        {
            var result = await runtime.RequestAsync<AgentStatus[]>("listAgents", new { });
            EnsureCurrent(runtime);
            Agents = result;
        }
        catch (Exception error)
        {
            if (reportError) ReportError(error);
            else throw;
        }
    }

    private async Task ReloadUsageAsync(RuntimeClient runtime, bool reset, bool reportError)
    {
        try
        {
            var query = CurrentUsageQuery(reset ? null : Usage.NextCursor, 20);
            var page = await runtime.RequestAsync<UsagePage>("queryUsage", new UsageParams(query));
            EnsureCurrent(runtime);
            Usage = reset ? page : page with { Items = [.. Usage.Items, .. page.Items] };
        }
        catch (Exception error)
        {
            if (reportError) ReportError(error);
            else throw;
        }
    }

    private UsageQuery CurrentUsageQuery(string? cursor, int? limit)
    {
        var now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        ulong? since = UsageRange switch
        {
            UsageRange.SevenDays => (ulong)Math.Max(0, now - 7 * 86_400),
            UsageRange.ThirtyDays => (ulong)Math.Max(0, now - 30 * 86_400),
            _ => null,
        };
        return new(UsageAgent, UsageModel, null, since, null, cursor, limit);
    }

    private async Task<GatewayState> RunStateAsync(RuntimeClient runtime, Task<GatewayState> operation)
    {
        try
        {
            IsBusy = true;
            var next = await operation;
            AcceptCurrent(runtime, next);
            return next;
        }
        catch (Exception error)
        {
            ReportError(error);
            throw;
        }
        finally { IsBusy = false; }
    }

    private void Accept(RuntimeClient source, GatewayState next)
    {
        if (ReferenceEquals(client, source) && source.IsAvailable && IsRuntimeAvailable) AcceptState(next);
    }

    private void AcceptCurrent(RuntimeClient source, GatewayState next)
    {
        EnsureCurrent(source);
        AcceptState(next);
    }

    private void AcceptState(GatewayState next)
    {
        var usageChanged = next.UsageRevision != State.UsageRevision;
        State = next;
        RaiseDerivedState();
        if (usageChanged && client is { } runtime && IsRuntimeAvailable) _ = ReloadUsageAsync(runtime, true, true);
    }

    private void RuntimeExited(RuntimeClient exited, Exception? error)
    {
        if (!ReferenceEquals(client, exited) || disposing) return;
        IsRuntimeAvailable = false;
        IsBusy = false;
        ClearRuntimeData();
        cleanupTask = exited.DisposeAsync().AsTask();
        Error?.Invoke(error?.Message ?? "The desktop runtime stopped.");
    }

    private void ClearRuntimeData()
    {
        State = GatewayState.Empty;
        Agents = [];
        Usage = UsagePage.Empty;
        ClientKey = "";
        RaiseDerivedState();
    }

    private RuntimeClient CurrentClient() => client is { } runtime && IsRuntimeAvailable
        ? runtime
        : throw new RuntimeException("runtime_unavailable", "The desktop runtime is not running. Restart it to continue.");

    private void EnsureCurrent(RuntimeClient source)
    {
        if (!ReferenceEquals(client, source) || !IsRuntimeAvailable)
            throw new RuntimeException("runtime_restarted", "The desktop runtime restarted before the operation completed.");
    }

    private void EnsureOwned(RuntimeClient source)
    {
        if (!ReferenceEquals(client, source))
            throw new RuntimeException("runtime_restarted", "The desktop runtime restarted before the operation completed.");
    }

    private void ReportError(Exception error)
    {
        if (!disposing) Error?.Invoke(error.Message);
    }

    private void RaiseDerivedState()
    {
        OnPropertyChanged(nameof(IsRunning));
        OnPropertyChanged(nameof(IsProtected));
        OnPropertyChanged(nameof(IsDevMode));
        OnPropertyChanged(nameof(ActiveProfile));
    }

    private void Set<T>(ref T field, T value, [CallerMemberName] string? name = null)
    {
        if (EqualityComparer<T>.Default.Equals(field, value)) return;
        field = value;
        OnPropertyChanged(name);
    }

    private void OnPropertyChanged(string? name) => PropertyChanged?.Invoke(this, new(name));

    public async ValueTask DisposeAsync()
    {
        disposing = true;
        await lifecycleLock.WaitAsync();
        try
        {
            IsRuntimeAvailable = false;
            await cleanupTask;
            if (client is { } runtime)
            {
                client = null;
                await runtime.DisposeAsync();
            }
        }
        finally { lifecycleLock.Release(); }
    }
}

public enum UsageRange { SevenDays, ThirtyDays, AllTime }
